source/octf/analytics/statistics/Distribution.cpp

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright(c) 2012-2018 Intel Corporation
+ * SPDX-License-Identifier: BSD-3-Clause-Clear
+ */
+
+#include <octf/analytics/statistics/Distribution.h>
+
+#include <cmath>
+#include <limits>
+#include <octf/utils/Exception.h>
+
+namespace octf {
+
+struct Distribution::Bucket {
+    Bucket()
+            : Id()
+            , Size()
+            , Power()
+            , Begin()
+            , End()
+            , Step()
+            , Sum()
+            , Count() {}
+
+    Bucket(const Bucket &other)
+            : Id(other.Id)
+            , Size(other.Size)
+            , Power(other.Power)
+            , Begin(other.Begin)
+            , End(other.End)
+            , Step(other.Step)
+            , Sum(other.Sum)
+            , Count(other.Count) {}
+
+    Bucket &operator=(const Bucket &other) {
+        if (&other != this) {
+            Id = other.Id;
+            Size = other.Size;
+            Power = other.Power;
+            Begin = other.Begin;
+            End = other.End;
+            Step = other.Step;
+            Sum = other.Sum;
+            Count = other.Count;
+        }
+
+        return *this;
+    }
+
+    Bucket(uint64_t id, uint64_t size, uint64_t power)
+            : Sum(size)
+            , Count(size) {
+        Id = id;
+        Size = size;
+        Power = power;
+        Begin = 0;
+        if (id) {
+            Begin = std::pow(power, id - 1) * size;
+        }
+        End = std::pow(power, id) * size - 1;
+        Step = (End - Begin + 1) / size;
+    }
+
+    void operator+=(uint64_t value) {
+        if (value < Begin || value > End) {
+            throw Exception("Updating wrong distribution bucket");
+        }
+
+        uint id = value;
+        id -= Begin;
+        id /= Step;
+
+        if (id < Size) {
+            Sum[id] += value;
+            Count[id]++;
+        } else {
+            throw Exception("Updating distribution bucket fatal error");
+        }
+    }
+
+    uint64_t Id;
```

Comment:
Is Id equivalent to Level? Please use same nomenclature as in header file

Reply:
It is. Name changed to level.